Sources/LanguageServerProtocolTransport/MessageSplitting.swift

```diff
@@ -10,14 +10,117 @@
 //
 //===----------------------------------------------------------------------===//
 
-import Foundation
+package import Foundation
 import LanguageServerProtocol
+@_spi(SourceKitLSP) import SKLogging
 
+/// A stateful JSONRPC message parser. Expects data to be read in chunks of `nextReadLength`.
+package class JSONMessageParser<MessageType> {
+  /// Decoder of the JSON message contents
+  private let decoder: (Data) -> MessageType?
+
+  /// Buffer of received bytes that haven't been fully parsed.
+  private nonisolated(unsafe) var requestBuffer: [UInt8] = []
+
+  /// Current parser state
+  private var state: ReadState = .header
+
+  /// The number of bytes to read given the current state of the parser.
+  package var nextReadLength: Int {
+    switch state {
+    case .header:
+      // Only ever read a single byte for the header to better handle invalid cases
+      return 1
+    case .message(let remaining):
+      // Up until the message, where we should read its entire length (or anything remaining if we had a partial read)
+      return remaining
+    }
+  }
+
+  package init(decoder: @escaping (Data) -> MessageType?) {
+    self.decoder = decoder
+  }
+
+  /// Parse the next chunk of data (see `nextReadLength`). Returns a message parsed by `decoder` if read and `nil`
+  /// otherwise (including on error). Note that this does not handle being passed a chunk that crosses header+contents
+  /// boundaries or whole message boundaries.
+  package func parse(chunk: Data) -> MessageType? {
+    switch state {
+    case .header:
+      parseHeader(data: chunk)
+      return nil
+    case .message(let remaining):
+      return parseMessage(data: chunk, remaining: remaining)
+    }
+  }
+
+  func parseHeader(data: Data) {
```

Comment:
```suggestion
  private func parseHeader(data: Data) {
```

I think for some other methods in here as well.